python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py

```diff
@@ -0,0 +1,273 @@
+    ParentMaterializedAutoMaterializeCondition,
+    ParentOutdatedAutoMaterializeCondition,
+)
+
+if TYPE_CHECKING:
+    from dagster._core.definitions.asset_daemon_context import AssetDaemonContext
+    from dagster._core.definitions.asset_daemon_cursor import AssetDaemonCursor
+
+
+class RuleEvaluationContext(NamedTuple):
+    asset_key: AssetKey
+    cursor: "AssetDaemonCursor"
+    instance_queryer: CachingInstanceQueryer
+    data_time_resolver: CachingDataTimeResolver
+    will_materialize_mapping: Mapping[AssetKey, AbstractSet[AssetKeyPartitionKey]]
+    expected_data_time_mapping: Mapping[AssetKey, Optional[datetime.datetime]]
+    candidates: AbstractSet[AssetKeyPartitionKey]
+    daemon_context: "AssetDaemonContext"
+
+    @property
+    def asset_graph(self) -> AssetGraph:
+        return self.instance_queryer.asset_graph
+
+    def materializable_in_same_run(self, child_key: AssetKey, parent_key: AssetKey) -> bool:
+        """Returns whether a child asset can be materialized in the same run as a parent asset."""
+        from dagster._core.definitions.external_asset_graph import ExternalAssetGraph
+
+        return (
+            # both assets must be materializable
+            child_key in self.asset_graph.materializable_asset_keys
+            and parent_key in self.asset_graph.materializable_asset_keys
+            # the parent must have the same partitioning
+            and self.asset_graph.have_same_partitioning(child_key, parent_key)
+            # the parent must have a simple partition mapping to the child
+            and (
+                not self.asset_graph.is_partitioned(parent_key)
+                or isinstance(
+                    self.asset_graph.get_partition_mapping(child_key, parent_key),
+                    (TimeWindowPartitionMapping, IdentityPartitionMapping),
+                )
+            )
+            # the parent must be in the same repository to be materialized alongside the candidate
+            and (
+                not isinstance(self.asset_graph, ExternalAssetGraph)
+                or self.asset_graph.get_repository_handle(child_key)
+                == self.asset_graph.get_repository_handle(parent_key)
+            )
+        )
+
+
+class AutoMaterializeRule(ABC):
+    @staticmethod
+    def materialize_on_required_for_freshness() -> "MaterializeOnRequiredForFreshnessRule":
+        return MaterializeOnRequiredForFreshnessRule()
+
+    @staticmethod
+    def materialize_on_parent_updated() -> "MaterializeOnParentUpdatedRule":
+        return MaterializeOnParentUpdatedRule()
+
+    @staticmethod
+    def materialize_on_missing() -> "MaterializeOnMissingRule":
+        return MaterializeOnMissingRule()
+
+    @staticmethod
+    def skip_on_parent_outdated() -> "SkipOnParentOutdatedRule":
+        return SkipOnParentOutdatedRule()
+
+    @abstractproperty
+    def decision_type(self) -> AutoMaterializeDecisionType:
+        ...
+
+    @abstractmethod
+    def evaluate(
```

Comment:
This evaluates a single asset, right? If so, I think we should include `asset` in the name. Some docstring would alse be helpful.